File: Android/sdk-weather-extension-android/weatherexample/src/main/java/com/adobe/marketing/mobile/weatherexample/WeatherExtensionListener.java
/*
 ADOBE CONFIDENTIAL

 Copyright 2019 Adobe
 All Rights Reserved.

 NOTICE: Adobe permits you to use, modify, and distribute this file in
 accordance with the terms of the Adobe license agreement accompanying
 it. If you have received this file from a source other than Adobe,
 then your use, modification, or distribution of it requires the prior
 written permission of Adobe.
 */

package com.adobe.marketing.mobile.weatherexample;

import com.adobe.marketing.mobile.Event;
import com.adobe.marketing.mobile.ExtensionApi;
import com.adobe.marketing.mobile.ExtensionListener;

class WeatherExtensionListener extends ExtensionListener {
    protected WeatherExtensionListener(final ExtensionApi extension, final String type, final String source) {
        super(extension, type, source);
    }

    @Override
    public void hear(final Event event) {
        if (event.getEventData() == null) {
            return;
        }

        final WeatherExtensionInternal parentExtension = getParentExtension();
        if (parentExtension == null) {
            return;
        }

        // handle SharedState events
        if (WeatherExtensionConstants.EVENT_TYPE_ADOBE_HUB.equalsIgnoreCase(event.getType())) {
            if (WeatherExtensionConstants.SharedState.CONFIGURATION.
                    equalsIgnoreCase((String)event.getEventData().get(WeatherExtensionConstants.SharedState.STATE_OWNER))) {
                getParentExtension().getExecutor().execute(new Runnable() {
                    @Override
                    public void run() {
                        parentExtension.processEvents();
                    }
                });
            }
        } else if (WeatherExtensionConstants.EVENT_TYPE_WEATHER_EXTENSION.equalsIgnoreCase(event.getType())) {
            // handle WeatherExtension events
            getParentExtension().getExecutor().execute(new Runnable() {
                @Override
                public void run() {
                    parentExtension.queueEvent(event);
                    parentExtension.processEvents();
                }
            });
        } else if (WeatherExtensionConstants.EVENT_TYPE_ADOBE_RULES_ENGINE.equalsIgnoreCase(event.getType()) &&
                WeatherExtensionConstants.EVENT_SOURCE_ADOBE_RESPONSE_CONTENT.equalsIgnoreCase(event.getSource())) {
            // handle Rules Consequences
            getParentExtension().getExecutor().execute(new Runnable() {
                @Override
                public void run() {
                    parentExtension.queueEvent(event);
                    parentExtension.processEvents();
                }
            });
        }
    }

    @Override
    protected WeatherExtensionInternal getParentExtension() {
        return (WeatherExtensionInternal) super.getParentExtension();
    }
}

File: iOS/ACPWeatherExample/ACPWeatherExample/WeatherExtensionListener.m
/*
 ADOBE CONFIDENTIAL
 
 Copyright 2019 Adobe
 All Rights Reserved.
 
 NOTICE: Adobe permits you to use, modify, and distribute this file in
 accordance with the terms of the Adobe license agreement accompanying
 it. If you have received this file from a source other than Adobe,
 then your use, modification, or distribution of it requires the prior
 written permission of Adobe.
 */

#import "ACPCore.h"
#import "WeatherExtensionInternal.h"
#import "WeatherExtensionListener.h"

@implementation WeatherExtensionListener

- (void) hear:(ACPExtensionEvent *)event {
    [ACPCore log:ACPMobileLogLevelDebug tag:@"WeatherExtensionListener" message:[NSString stringWithFormat:@"WeatherExtensionListener heard an event: %@, %@.  Data: %@", event.eventName, event.eventType, event.eventData]];
    WeatherExtensionInternal* parentExtension = [self getParentExtension];
    if (parentExtension == nil) {
        [ACPCore log:ACPMobileLogLevelWarning tag:@"WeatherExtensionListener" message:@"The parent extension was nil, skipping event"];
        return;
    }

    // handle SharedState events
    if ([event.eventType isEqualToString:@"com.adobe.eventType.hub"]) {
        if ([event.eventData[@"stateowner"] isEqualToString:@"com.adobe.module.configuration"]) {
            [parentExtension processEvents];
        }
    }

    // handle WeatherExtension events
    else if ([event.eventType isEqualToString:@"com.acpExample.eventType.weatherExtension"]) {
        [parentExtension queueEvent:event];
        [parentExtension processEvents];
    }
    
    // handle Rules Consequences
    else if ([event.eventType isEqualToString:@"com.adobe.eventType.rulesEngine"] && [event.eventSource isEqualToString:@"com.adobe.eventSource.responseContent"]) {
        // todo: check consequence format
        [parentExtension queueEvent:event];
        [parentExtension processEvents];
    }
}

/**
 * Returns the parent extension that owns this listener
 */
- (WeatherExtensionInternal*) getParentExtension {
    WeatherExtensionInternal* parentExtension = nil;
    if ([[self extension] isKindOfClass:WeatherExtensionInternal.class]) {
        parentExtension = (WeatherExtensionInternal*) [self extension];
    }

    return parentExtension;
}

@end
